### PR TITLE
feat: introduce pgoutput plugin

### PR DESCRIPTION
**Pull request by KennyChenFight**
This is a draft version for support pgoutput plugin when pg version >= 14.

1. implement pgoutput plugin decoder (only support binary format)
 	* ref message format: https://github.com/postgres/postgres/blob/master/src/backend/replication/logical/proto.c
 	* ref pgoutput decode: https://github.com/postgres/postgres/blob/master/src/backend/replication/pgoutput/pgoutput.c
 	* separate pglogical and pgoutput decoder implementation
3. setup ci for pg14
	* setup pg14 logical docker image and use pglogical patch to remove filter and local node
	* use PG_VERSION env to decide should run pgoutput test or not

### CODE REVIEW

**Comment by rueian** [comment]
This is really cool! Thanks @KennyChenFight!

BTW, is there any chance to drop the pglogical dependency?

**Comment by KennyChenFight** [comment]
> This is really cool! Thanks @KennyChenFight!
> 
> BTW, is there any chance to drop the pglogical dependency?

Now, the client has the option to choose which decoder to use. By default, the pglogical decoder is used. However, if the client is using PostgreSQL version 14 or above, they can specify to use the pgoutput decoder instead. In this case, the client does not need to install pglogical.
So, we still can keep pglogical decoder implemation in code ?

**Comment by rueian** [comment]
Sure, we should keep the pglogical implementation. However, I believe making pgoutput as default is better since it lowers the barrier to use this library for new users.

**Comment by KennyChenFight** [comment]
> Sure, we should keep the pglogical implementation. However, I believe making pgoutput as default is better since it lowers the barrier to use this library for new users.

Considering that the pgoutput binary format is only supported in PG 14 or above, and PG 14 is still a relatively newer version, it might be better to set the pglogical decoder as the default option. Or, maybe I could implement pgoutput text format ?

File: pkg/sql/source.go [hunk 1]
```diff
@@ -19,6 +19,8 @@ WHERE (i.indisprimary OR i.indisunique) AND i.indisvalid AND i.indpred IS NULL O
 
 var CreateLogicalSlot = `SELECT pg_create_logical_replication_slot($1, $2);`
 
+var CreatePublication = `CREATE PUBLICATION %s FOR ALL TABLES;`
```
Comment (rueian) [inline pkg/sql/source.go:22]: Why do we need this? Isn't this meant to be used with `CREATE SUBSCRIPTION`?

File: pkg/decode/bytes.go [hunk 2]
```diff
@@ -63,6 +64,10 @@ func (b *BytesReader) Int8() (v int, err error) {
 	return int(uv), err
 }
 
+func (b *BytesReader) IntEndIdx() (v int) {
+	return bytes.IndexByte(b.data[b.off:], byte(0))
```
Comment (rueian) [inline pkg/decode/bytes.go:68]: Should we consider returning an error if we can't find a byte(0)?

File: pkg/sql/source.go (hunk 1 above)
Comment (KennyChenFight) [reply pkg/sql/source.go:22]: publication is more like pglogical local node provider.
Each publication can be matched a replication slot and a subscription, but we only need publication and replication slot.
When starting replication, we should use pgoutput plugin and need to give publication names as plugin parameters.

File: pkg/decode/bytes.go (hunk 2 above)
Comment (KennyChenFight) [reply pkg/decode/bytes.go]: done

**Review by rueian — APPROVED** [review]

**Pull request merged**